src/serde/mod.rs

```diff
@@ -0,0 +1,271 @@
+//!     // data structure for representing address
+//!     let address = Address {
+//!         street: "10 Downing Street".to_owned(),
+//!         city: "London".to_owned(),
+//!     };
+//!
+//!     // serialize it to an Ion data
+//!     let ion = to_string(&address)?;
+//!
+//!     // assert that the serialized Ion data is as expected
+//!     assert_eq!(r#"{street: "10 Downing Street", city: "London"}"#, ion);
+//!
+//!     Ok(())
+//! }
+//! ```
+//!
+//! ## Example of deserialization of Ion data into Rust struct
+//! ```
+//! use ion_rs::IonResult;
+//! use crate::ion_rs::serde::from_ion;
+//! use serde::{Deserialize, Serialize};
+//!
+//!#[derive(Serialize, Deserialize)]
+//! struct Address {
+//!     street: String,
+//!     city: String,
+//! }
+//!
+//! fn main() -> IonResult<()> {
+//!     // represents Ion data with address information
+//!     let data = r#"
+//!         {
+//!             street: "10 Downing Street",
+//!             city: "London"    
+//!         }
+//!     "#;
+//!
+//!     // deserialize Ion data into Rust struct for address
+//!     let address: Address = from_ion(data)?;
+//!
+//!     // assert that the deserialized Rust struct has street and city field set correctly
+//!     assert_eq!(address.street, "10 Downing Street");
+//!     assert_eq!(address.city, "London");
+//!
+//!     Ok(())
+//! }
+//! ```
+//!
+//! ## Example of serialization and deserialization for Timestamp
+//!```
+//! use serde::{Deserialize, Serialize};
+//! use ion_rs::Timestamp;
+//! use ion_rs::IonResult;
+//! use ion_rs::serde::from_ion;
+//! use serde_with::serde_as;
+//! use chrono::{Utc, TimeZone, FixedOffset, DateTime};
+//!
+//! #[serde_as]
+//! #[derive(Serialize, Deserialize)]
+//! struct Event {
+//!     name: String,
+//!     start_time: Timestamp,
+//!     #[serde_as(as = "crate::Timestamp")]
+//!     end_time: DateTime<FixedOffset>
+//! }
+//!
+//! fn main() -> IonResult<()> {
+//! // represents Ion data with event information
+//! let data = r#"
+//!         {
+//!             name: "Annual Conference",
+//!             start_time: 2023-01-01T16:30:00Z,
+//!             end_time: 2023-01-01T18:00:00Z
+//!         }
+//!     "#;
+//!
+//!     // deserialize Ion data into Rust struct for event
+//!     let event: Event = from_ion(data)?;
+//!
+//!     // assert that the deserialized Rust struct has name, start_time and end_time set correctly
+//!     assert_eq!(event.name, "Annual Conference");
+//!     assert_eq!(event.start_time, Timestamp::with_ymd(2023, 1, 1).with_hms(16, 30, 0).build()?);
+//!     assert_eq!(event.end_time, Utc.with_ymd_and_hms(2023, 1, 1, 18, 0, 0).unwrap());
+//!
+//!    Ok(())
+//! }
+//! ```
+//!
+//! ## Example of serialization and deserialization for Decimal
+//!```
+//! use serde::{Deserialize, Serialize};
+//! use ion_rs::Decimal;
+//! use ion_rs::IonResult;
+//! use ion_rs::serde::from_ion;
+//!
+//! #[derive(Serialize, Deserialize)]
+//! struct Product {
+//!     name: String,
```

Comment:
It'd be good to have an example that shows deserializing strings and byte arrays to a `&'a str` and `&'a [u8]` respectively.